network/p2p/subscription_filter.go

```diff
@@ -0,0 +1,71 @@
+package p2p
+
+import (
+	"github.com/libp2p/go-libp2p-core/peer"
+	pb "github.com/libp2p/go-libp2p-pubsub/pb"
+
+	"github.com/onflow/flow-go/engine"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/id"
+	"github.com/onflow/flow-go/network"
+)
+
+// RoleBasedFilter implements a subscription filter that filters subscriptions based on a node's role.
+type RoleBasedFilter struct {
+	idProvider  id.IdentityProvider
+	myPeerID    peer.ID
+	rootBlockID flow.Identifier
+	chainID     flow.ChainID
+}
+
+func NewRoleBasedFilter(pid peer.ID, rootBlockID flow.Identifier, chainID flow.ChainID, idProvider id.IdentityProvider) *RoleBasedFilter {
+	return &RoleBasedFilter{
+		idProvider,
+		pid,
+		rootBlockID,
+		chainID,
+	}
+}
+
+func (f *RoleBasedFilter) allowedTopics(pid peer.ID) map[network.Topic]struct{} {
```

Comment:
shouldn't we have two different filter. One which all the staked node use and one which the full observer node uses and the Staked AN that support the observer node use?

Reply:
We could do this, the at a high level, the logic for the subscription filter goes:

A node should only be allowed to subscribe to some set of channels based on its role.

This general rule is the same for any node type, so we can implement the exact same logic for all of them